src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -0,0 +1,183 @@
+                var variableState = new VariableState(globalState.VariableSlot, globalState.VariableBySlot, globalState.VariableTypes, incrementalSnapshot.VariableState.Clone());
+                var method = globalState.Symbol as MethodSymbol;
+                return new NullableWalker(
+                    compilation,
+                    globalState.Symbol,
+                    useMethodSignatureParameterTypes: !(method is null),
+                    method,
+                    nodeToAnalyze,
+                    binder,
+                    binder.Conversions,
+                    variableState,
+                    returnTypesOpt: null,
+                    analyzedNullabilityMap,
+                    snapshotBuilderOpt: null);
+            }
+
+#if DEBUG
+            internal void VerifyNode(BoundNode node)
+            {
+                if (node.Kind == BoundKind.TypeExpression || node.WasCompilerGenerated)
+                {
+                    return;
+                }
+
+                Debug.Assert(_incrementalSnapshots.ContainsKey(node.Syntax.SpanStart), $"Did not find a snapshot for {node} `{node.Syntax}.`");
+                Debug.Assert(_walkerGlobalStates.Length > _incrementalSnapshots[node.Syntax.SpanStart].GlobalStateIndex, $"Did not find global state for {node} `{node.Syntax}`.");
+            }
+#endif
+
+            internal sealed class Builder
+            {
+                private readonly ArrayBuilder<SharedWalkerState> _walkerStates = ArrayBuilder<SharedWalkerState>.GetInstance();
+                private readonly ImmutableDictionary<int, Snapshot>.Builder _incrementalSnapshots = ImmutableDictionary.CreateBuilder<int, Snapshot>();
+                private readonly PooledDictionary<Symbol, int> _symbolToSlot = PooledDictionary<Symbol, int>.GetInstance();
+                private readonly ArrayBuilder<int> _previousWalkerSlots = ArrayBuilder<int>.GetInstance();
+                private int _currentWalkerSlot = -1;
+                private int _nextWalkerSlot = 0;
+
+                internal SnapshotManager ToManagerAndFree()
+                {
+                    Debug.Assert(_currentWalkerSlot == -1 && _previousWalkerSlots.Count == 0, "Attempting to finalize snapshots before all walks completed");
+                    Debug.Assert(_walkerStates.Count == _nextWalkerSlot);
+                    Debug.Assert(_symbolToSlot.Count == _walkerStates.Count);
+                    _symbolToSlot.Free();
+                    _previousWalkerSlots.Free();
+                    return new SnapshotManager(
+                        _walkerStates.ToImmutableAndFree(),
+                        _incrementalSnapshots.ToImmutable());
+                }
+
+                internal void EnterNewWalker(Symbol symbol)
+                {
+                    Debug.Assert(!(symbol is null));
+                    _previousWalkerSlots.Push(_currentWalkerSlot);
```

Comment:
>_previousWalkerSlots.Push(_currentWalkerSlot); [](start = 20, length = 46)

We can avoid allocating a special stack for that by simply returning previous slot to the caller and passing it back into ExitWalker.

Reply:
We could, but I would really prefer to keep the knowledge of the walker slots encapuslated within the builder and manager. I'll change it if you insist, but that's a code smell to me.

---
In reply to: [289483513](https://github.com/dotnet/roslyn/pull/35850#discussion_r289483513) [](ancestors = 289483513)

Reply:
I agree with @AlekseyTs - the previous symbol should be returned to the caller.

---
In reply to: [289492447](https://github.com/dotnet/roslyn/pull/35850#discussion_r289492447) [](ancestors = 289492447,289483513)